lib/Dialect/Arc/Transforms/FindInitialVectors.cpp

```diff
@@ -0,0 +1,222 @@
+    llvm::raw_string_ostream os(str);
+    elem.print(os);
+    key += str;
+  }
+  return success();
+}
+
+std::string computeKey(Operation *op) {
+  // The key = concat(op_rank, op_name, op_operands_types, op_result_types, op_attrs)
+  std::string key = op->getName().getStringRef().str();
+  if (failed(modifyKeyIfPossible(op->getOperandTypes(), key)) ||
+      failed(modifyKeyIfPossible(op->getResultTypes(), key)) || 
+      failed(modifyKeyIfPossible(op->getRegisteredInfo()->getAttributeNames(), key)))
+    assert(false && "Cannot compute the op key");
+
+  return key;
+}
+
+struct Vectorizer {
+  LogicalResult collectSeeds(Block *block) {
+    if (failed(order.compute(block)))
+      return failure();
+
+    for(auto &[op, rank] : order.opRanks)
+      candidates[std::to_string(rank) + computeKey(op)].push_back(op);
+
+    return success();
+  }
+
+  LogicalResult vectorize();
+  // Store Isomorphic ops together
+  std::unordered_map<std::string, SmallVector<Operation *>> candidates;
+  TopologicalOrder order;
+};
+} // namespace
+
+// When calling this function we assume that we have the candidate groups of
+// isomorphic ops so we need to feed them to the `VectorizeOp`
+LogicalResult Vectorizer::vectorize() {
+  // Unachievable?! just in case!
+  if (candidates.empty())
+    return success();
+
+  // Iterate over every group of isomorphic ops  
+  for (const auto &[key, ops] : candidates) {
+    // If the group has only one scalar then it doesn't worth vectorizing
+    if (ops.size() == 1)
+      continue;
+
+    // Here, we have a bunch of isomorphic ops, we need to extract the operands
+    // results and attributes of every op and store them in a vector
+    // Holds the operands
+    SmallVector<ValueRange> operandsVector;
+    // Holds the results
+    SmallVector<Type> resultTypes(ops.size(), ops[0]->getResult(0).getType());
+    for (const auto &op : ops) 
+      operandsVector.push_back(op->getOperands());
+
+    // Now construct the `VectorizeOp`
+    ImplicitLocOpBuilder builder(ops[0]->getLoc(), ops[0]);
+    auto vectorizeOp = builder.create<VectorizeOp>(resultTypes, operandsVector);
+    vectorizeOp->setAttrs(ops[0]->getAttrs());
+    // Now we have the operands, results and attributes, now we need to get
+    // the blocks.
+
+    // There was no blocks so we need to create one and set the insertion point
+    // at the first of this region
+    auto &vectorizeBlock = vectorizeOp.getBody().emplaceBlock();
+    builder.setInsertionPointToStart(&vectorizeBlock);
+
+    // Add the block arguments 
+    // comb.and %x, %y
+    // comb.and %u, %v
+    // then the operands' vector will be {{x, y}, {u, v}}
+    // we need to create an th block args, so we need the type and the location
+    // the type is a vectpr type
+    for (auto operands : operandsVector) {
+      mlir::Type vectorType = mlir::VectorType::get({-1}, operands.front().getType());
+      auto arg = vectorizeBlock.addArgument(vectorType, ops[0]->getLoc());
```

Comment:
@fabianschuiki After this line I think I have the block arguments. but how to bind the block args with the outer variables(ops).

```
%0 = comb.add %x, %y : i4
%1 = comb.add %a, %b : i4
```
after this line I think I will have two vectors {x, y} and {a, b} and the argument type for the `arc.vectorize` block.
```
%2 = arc.vectorize(%x, %y), (%a, %b) : (i4, i4, i4, i4) -> (i4, i4) {
^bb0(%arg0: vector<2xi4>, %arg1: vector<2xi4>) {
   %3 = comb.add %arg0, %arg1 // here, how to bind the vectors to arg0, and arg1??
   arc.output %3: vector<2xi4> 
}
}
```
Is there anything wrong with this?

Reply:
Are you sure you have vectors `{x, y}` and `{a, b}`? Looking at your code snippet, `x` and `a` appear on the LHS of the `comb.add`, and `y` and `b` appear on the RHS. If you collapse the two adds into a vectorized version, you'd want to have vector `{x, a}` on the LHS and `{y, b}` on the RHS.

In the `arc.vectorize` op, your block arguments don't need to have a vector type. Instead, the body of `arc.vectorize` is a scalar description of what should happen on _each_ vector lane. Something like this:
```mlir
%0, %1 = arc.vectorize (%x, %a), (%y, %b) : (i4, i4, i4, i4) -> (i4, i4) {
^bb0(%arg0: i4, %arg1: i4):
   %2 = comb.add %arg0, %arg1 : i4
   arc.vectorize.return %2 : i4
}
```

The vectorize op performs a "zip" over the operand groups and feeds each tuple of operands, `(x, y)` and `(a, b)` in your case, to its body. The body performs its scalar operation and returns the scalar result. The vectorize op then returns all these results. It's more or less like a `map` operation, where the scalar body is applied to each of the vector lanes.

This has the advantage of allowing us to vectorize the operands flowing into and results flowing out of the vectorize op at a later point. And, independently, to vectorize the body at a later point. That decouples forming the vectors from performing any kind of lowering or vectorization of the actual ops.